### PR TITLE
feat: prerender & analyse in worker rather than subprocess to support Deno

### PR DESCRIPTION
**Pull request by fernandolguevara**
use `workers` instead of `child_process`

related:
https://github.com/denoland/deno/issues/17248

@benmccann would you be able to review this PR?

### CODE REVIEW

File: packages/kit/src/utils/fork.js [hunk 1]
```diff
@@ -36,34 +34,31 @@ export function forked(module, callback) {
...
-				serialization: 'advanced'
+					...process.env
+				}
 			});
 
-			child.on(
+			worker.on(
 				'message',
-				/** @param {any} data */ (data) => {
+				/** @param {any} data */ async (data) => {
```
Comment (Rich-Harris) [inline packages/kit/src/utils/fork.js:45]: I think this can still be sync?

```suggestion
				/** @param {any} data */ (data) => {
```

**Review by Rich-Harris — APPROVED** [review]
this LGTM! thank you. just had one small observation, inline.

I'm not sure I totally understand how this affects Deno though — it looks like `child_process` and `worker_threads` are both implemented. Is there some nuance I'm missing?

- https://deno.land/std@0.142.0/node/child_process.ts
- https://deno.land/std@0.142.0/node/worker_threads.ts

**Comment by benmccann** [comment]
> I'm not sure I totally understand how this affects Deno though — it looks like child_process and worker_threads are both implemented. Is there some nuance I'm missing?

It sounds to me like Deno implements `child_process`, but not communication between child and parent process and for that recommend using workers - or at least that's what I got from [this article](https://medium.com/deno-the-complete-reference/going-from-node-js-to-deno-part-5-child-process-bde9cd21d3d6). @bartlomieju would you be able to confirm if that's correct?

**Comment by bartlomieju** [comment]
Thanks for the ping @benmccann. 

> It sounds to me like Deno implements child_process, but not communication between child and parent process 

That's true, we currently don't have support for the `child_process.send()` API. Unfortunately our polyfill for the `worker_threads` module is broken at the moment, but given that you have a need for it, we can prioritize fixing it and get it done next week. Does that sounds reasonable to you?

**Comment by benmccann** [comment]
Thanks! Yeah, we can wait to merge this on our side then. I'm also fine going either the `worker_threads` or `child_process.send()` route if one is easier than the other

**Comment by bartlomieju** [comment]
Workers would definitely be easier - we just need to update once-working code. I'll prioritize this work over the next week and get back to you once it's released (I'll target to release in Deno v1.33.4 next week).

**Comment by dominikg** [comment]
iirc we have been using child_process to ensure resources are properly cleaned up (eg db connections). Is that still the case with worker?

**Comment by bartlomieju** [comment]
> iirc we have been using child_process to ensure resources are properly cleaned up (eg db connections). Is that still the case with worker?

@dominikg I believe so - workers use separate OS threads, so once the worked is closed/terminated the resources will be released back to the OS.

**Comment by Rich-Harris** [comment]
This code guarantees that everything is correctly shut down — if it wasn't, the tests would never complete:

https://github.com/sveltejs/kit/blob/9f5642d156a797160abac58a29beca417bf53df8/packages/kit/test/prerendering/basics/src/hooks.server.js#L20-L23

So given that the tests are passing I'm confident that it works correctly.

**Comment by pboguslawski** [comment]
Please take a look at https://github.com/sveltejs/kit/issues/9989 - seems that using workers require also env cleanup.

**Pull request merged**